font/src/ft/mod.rs

```diff
@@ -529,6 +581,64 @@ impl FreeTypeRasterizer {
     }
 }
 
+fn downsample_bitmap(mut bitmap_glyph: RasterizedGlyph, fixup_factor: f64) -> RasterizedGlyph {
+    // Only scale colored buffers which are bigger than required
+    let bitmap_buffer = match (&bitmap_glyph.buf, fixup_factor.partial_cmp(&1.0)) {
+        (BitmapBuffer::RGBA(buffer), Some(Ordering::Less)) => buffer,
+        _ => return bitmap_glyph,
+    };
+
+    let bitmap_width = bitmap_glyph.width as usize;
+    let bitmap_height = bitmap_glyph.height as usize;
+
+    let target_width = (bitmap_width as f64 * fixup_factor) as usize;
+    let target_height = (bitmap_height as f64 * fixup_factor) as usize;
+
+    let downsampling_step = 1.0 / fixup_factor;
+
+    let mut downsampled_buffer = Vec::<u8>::with_capacity(target_width * target_height * 4);
+
+    for line_index in 0..target_height {
+        let line_index = line_index as f64;
+        let source_line_start = (line_index * downsampling_step).round() as usize;
+        let source_line_end = ((line_index + 1.) * downsampling_step).round() as usize;
+
+        for column_index in 0..target_width {
+            let column_index = column_index as f64;
+            let source_column_start = (column_index * downsampling_step).round() as usize;
+            let source_column_end = ((column_index + 1.) * downsampling_step).round() as usize;
+
+            let (mut r, mut g, mut b, mut a) = (0u32, 0u32, 0u32, 0u32);
+            let mut pixels_picked: u32 = 0;
+
+            for source_line in source_line_start..source_line_end {
+                let source_pixel_index = source_line * bitmap_width;
+
+                for source_column in source_column_start..source_column_end {
+                    let offset = (source_pixel_index + source_column) * 4;
+                    r += bitmap_buffer[offset] as u32;
+                    g += bitmap_buffer[offset + 1] as u32;
+                    b += bitmap_buffer[offset + 2] as u32;
+                    a += bitmap_buffer[offset + 3] as u32;
+                    pixels_picked += 1;
+                }
+            }
+
+            downsampled_buffer.push((r / pixels_picked) as u8);
+            downsampled_buffer.push((g / pixels_picked) as u8);
+            downsampled_buffer.push((b / pixels_picked) as u8);
+            downsampled_buffer.push((a / pixels_picked) as u8);
+        }
+    }
+
+    bitmap_glyph.top = (bitmap_glyph.top as f64 * fixup_factor) as i32;
+    bitmap_glyph.left = (bitmap_glyph.left as f64 * fixup_factor) as i32;
+    bitmap_glyph.width = target_width as i32;
+    bitmap_glyph.height = target_height as i32;
+    bitmap_glyph.buf = BitmapBuffer::RGBA(downsampled_buffer);
+    bitmap_glyph
```

Comment:
I've looked at this for every review I've done so far and I never really enjoyed how big of a 'blob' of code this is. However I never really wanted to say anything because I feel like this does all fit together nicely and we of course don't want to just add useless newlines.

However I'd suggest the following change, which I think improves things a bit:

```rust
bitmap_glyph.buf = BitmapBuffer::RGBA(downsampled_buffer);

// Downscale the metrics
bitmap_glyph.top = (bitmap_glyph.top as f64 * fixup_factor) as i32;
bitmap_glyph.left = (bitmap_glyph.left as f64 * fixup_factor) as i32;
bitmap_glyph.width = target_width as i32;
bitmap_glyph.height = target_height as i32;

bitmap_glyph
```

I think putting the assignment to the buffer at the front is nice because that's mostly what we've been working on until now. So it's a bit more tied to the loop we've been going through.

Having the metrics in a separate block allows us to add a nice comment, making it easier to read (or making it possible not to read it at all!).

Having the return on a separate line is always nice I think because it just separates it a bit and makes it clear that we actually really want to return something and haven't just forgotten a semicolon.